docs/proof_object.md

```diff
@@ -275,3 +275,91 @@ proof checker extremely simple.
 * Check that `P11` is syntactically equal to `Q1`
 * Check that `P12` is syntactically equal to `P21`
 * Check that `P22` is syntactically equal to `Q2`
+
+# Mixed-level Proof Object Syntax
+
+Language definitions are expected to use mostly
+object-level or mixed-level formulas, which is why
+we have defined the Kore syntax to allow these formulas,
+rather than simply requiring everything to be encoded
+in meta-level formulas.
+For the same reason, we wish to define a syntax of proofs
+that allows writing mixed-level formulas, and that allows
+object-level reasoning to be written in the
+form `by <ProofRule>` rather than by explicitly
+constructing meta-level proofs of terms involving `#provable`
+using the hypotheses described in section 7.7 of
+The Semantics of K.
+
+To support this we allow as patterns any Kore pattern,
+according to the syntax of `<pattern>` in 9.1.4.
+```
+  <ProofObject>  ::= <AssumedClaim>* <DerivedClaim>*
+
+  <AssumedClaim> ::= <Id> : <Pattern>
+
+  <DerivedClaim> ::= <Id> : <Pattern> by <ProofRule>
+
+  <Id> ::= /* Quite flexible syntax. */
+```
+
+Proofs will be defined so that a valid mixed-level proof can
+be transformed into a valid simple proof that assumes or
+proves (resp.) the lifting of each assumed or derived claim
+of the mixed-level proof.
+
+#### Note
+We may wish to disambiguate the object-level proof rule
+
+A pattern is considered equivalent to its lifting.
+Following the convention of lifting we will allow some
+proof rule names to be re-used between object-level and meta-level
+forms, prefixing a '#' to unambiguously refer to the meta-level form.
+
+All the proof rules above should be available at the object level as well,
+but now the actual reasoning will be in the form of properly
+instantiating and applying a corresponding meta-level hypothesis.
+
+The "object-level" form of the proof rules are equivalent to proving
+`#provable` of the lifting of the formula.
+To exactly resemble the meta-level proof rules, the `#wellFormed`
+hypotheses are automatically discharged.
+
+The new proof rule `check-well-formed(P)` proves
+that `#wellFormed` holds of the lifting of P whenever
+P is well-formed.
+
+Similarly, hypotheses of the form `#occursFree{s}(x,P)`
+or `\not{s}(#occursFree{s}(x,P))` are also automatically
+discharged, and new proof rules `check-free(x,P)`
+and `check-not-free(x,P)`
+
+```
+  <ProofRule>    ::=

```

Comment:
Maybe make it clear and say `<MixedLevelProofRule> ::= <ObjectLevelProofRule> | <MetaLevelProofRule>` and explicitly use `#propositional1`, `#propositional2`, etc., for the meta-level rules.

Reply:
Do you think we should not overload the rule names? If not, consider that it would be the prefixed names in a mixed proof that desugar directly to the (unprefixed) meta-level proof rules.

Reply:
I think we shouldn't overload rule names. In addition, I prefer we adopt a consistent naming convention here, that is `#...` for meta-level proof rules and non-`#...` for object-level proof rules.

Reply:
Do you think we should also add the `#` prefix to pure matching logic proofs, or just to meta-level rules in mixed proofs?